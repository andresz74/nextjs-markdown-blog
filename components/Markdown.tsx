import React from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import CodeBlock from '@/utils/markdownRenderers/CodeBlock';
import Heading from '@/utils/markdownRenderers/Heading';
import ImageNode from '@/utils/markdownRenderers/ImageNode'
import LinkNode from '@/utils/markdownRenderers/LinkNode';
import ListNode from '@/utils/markdownRenderers/ListNode';
import ListItemNode from '@/utils/markdownRenderers/ListItemNode';
import Paragraph from '@/utils/markdownRenderers/Paragraph';
import Preformatted from '@/utils/markdownRenderers/Preformatted';
import { Components } from 'react-markdown';

interface MarkdownProps {
	children: string;
}

const renderers: Components = {
	a: ({ children, ...props }) => <LinkNode {...props}>{children}</LinkNode>,
	code: ({ className, children, ...props }) => {
		// Block-level code: if `className` is present, it's a fenced code block
		if (className) {
			return <CodeBlock className={className}>{children}</CodeBlock>;
		}
		// Inline code: render without breaking the paragraph
		return <code style={{
			backgroundColor: '#f0f0f0',
			padding: '2px 4px',
			borderRadius: '4px',
			fontFamily: 'monospace',
			fontSize: '0.875rem',
		}} {...props}>{children}</code>;
	},
	h1: ({ children, ...props }) => <Heading level={1} {...props}>{children}</Heading>,
	h2: ({ children, ...props }) => <Heading level={2} {...props}>{children}</Heading>,
	h3: ({ children, ...props }) => <Heading level={3} {...props}>{children}</Heading>,
	h4: ({ children, ...props }) => <Heading level={4} {...props}>{children}</Heading>,
	h5: ({ children, ...props }) => <Heading level={5} {...props}>{children}</Heading>,
	h6: ({ children, ...props }) => <Heading level={6} {...props}>{children}</Heading>,
	img: ({ alt, src, ...props }) => <ImageNode alt={alt ?? ''} src={src ?? ''} {...props} />,
	li: ({ children, ...props }) => <ListItemNode {...props}>{children}</ListItemNode>,
	ol: ({ children, ...props }) => <ListNode ordered={true} {...props}>{children}</ListNode>,
	ul: ({ children, ...props }) => <ListNode ordered={false} {...props}>{children}</ListNode>,
	p: ({ children, ...props }) => <Paragraph {...props}>{children}</Paragraph>,
	pre: ({ children, ...props }) => <Preformatted {...props}>{children}</Preformatted>,
};

const Markdown: React.FC<MarkdownProps> = ({ children }) => {
	return (
		<ReactMarkdown components={renderers} remarkPlugins={[[remarkGfm, { singleTilde: false }]]}>
			{children}
		</ReactMarkdown>
	);
};

export default Markdown;
