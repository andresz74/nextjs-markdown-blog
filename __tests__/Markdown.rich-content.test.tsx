import React from 'react';
import { render, screen } from '@testing-library/react';
import Markdown from '@/components/Markdown';

describe('Markdown rich content rendering', () => {
	it('renders image', () => {
		const md = `![Alt text](/images/test.png)`;
		const { getByAltText } = render(<Markdown>{md}</Markdown>);
		const img = getByAltText('Alt text');
		expect(img).toBeInTheDocument();
		expect(img.getAttribute('src')).toContain('/images/test.png');
	});

	it('renders image with no alt or src', () => {
		render(<Markdown>{`![]()`}</Markdown>);
		const img = screen.getByRole('img');
		expect(img).toBeInTheDocument();
		expect(img).toHaveAttribute('alt', 'Image');
		expect(img).toHaveAttribute('src', '/default-image.jpg');
	});

	it('renders image with defined src and missing alt', () => {
		render(<Markdown>{`![](/images/defined-src.png)`}</Markdown>);
		const img = screen.getByRole('img');
		expect(img).toBeInTheDocument();
		expect(img).toHaveAttribute('alt', 'Image'); // fallback still triggered
		expect(img).toHaveAttribute('src', '/images/defined-src.png'); // defined src
	});

	it('renders image with defined alt and missing src', () => {
		render(<Markdown>{`![Alt text]()`}</Markdown>); // no src
		const img = screen.getByAltText('Alt text');
		expect(img).toBeInTheDocument();
		expect(img).toHaveAttribute('src', '/default-image.jpg'); // or '', depending on ImageNode fallback
	});

	it('renders code block', () => {
		const md = '```ts\nconst x: number = 42;\n```';
		const { container } = render(<Markdown>{md}</Markdown>);
		expect(container).toMatchSnapshot();
	});

	it('renders unordered list', () => {
		render(<Markdown>{`- Item 1\n- Item 2`}</Markdown>);
		expect(screen.getByText('Item 1')).toBeInTheDocument();
		expect(screen.getByText('Item 2')).toBeInTheDocument();
	});

	it('renders blockquote', () => {
		render(<Markdown>{'> This is a quote'}</Markdown>);
		expect(screen.getByText('This is a quote')).toBeInTheDocument();
	});

	it('renders inline code', () => {
		render(<Markdown>{'Text with `inline code`.'}</Markdown>);
		expect(screen.getByText('inline code')).toBeInTheDocument();
	});

	it('renders styled inline code', () => {
		const md = 'Some `code` inline.';
		const { container } = render(<Markdown>{md}</Markdown>);
		const codeEl = container.querySelector('code');

		expect(codeEl).toBeInTheDocument();
		expect(codeEl?.textContent).toBe('code');

		// Validate inline styles â€” trigger execution of every key in line 43
		expect(codeEl).toHaveStyle({
			backgroundColor: '#f0f0f0',
			padding: '2px 4px',
			borderRadius: '4px',
			fontFamily: 'monospace',
			fontSize: '0.875rem',
		});
	});

	it('renders headings correctly', () => {
		render(<Markdown>{`# H1\n## H2\n### H3\n#### H4\n##### H5\n###### H6`}</Markdown>);
		expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent('H1');
		expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent('H2');
		expect(screen.getByRole('heading', { level: 3 })).toHaveTextContent('H3');
		expect(screen.getByRole('heading', { level: 4 })).toHaveTextContent('H4');
		expect(screen.getByRole('heading', { level: 5 })).toHaveTextContent('H5');
		expect(screen.getByRole('heading', { level: 6 })).toHaveTextContent('H6');
	});

	it('renders a link correctly', () => {
		render(<Markdown>{`[OpenAI](https://openai.com)`}</Markdown>);
		const link = screen.getByRole('link', { name: 'OpenAI' });
		expect(link).toBeInTheDocument();
		expect(link).toHaveAttribute('href', 'https://openai.com');
	});

	it('renders plain paragraph text', () => {
		render(<Markdown>{`This is just plain text.`}</Markdown>);
		expect(screen.getByText('This is just plain text.')).toBeInTheDocument();
	});

	it('renders fallback for image with missing src', () => {
		render(<Markdown>{`![Alt text]()`}</Markdown>);
		const img = screen.getByAltText('Alt text');
		expect(img).toBeInTheDocument();
	});

	it('renders emojis correctly', () => {
		render(<Markdown>{`Hello ðŸ‘‹ world!`}</Markdown>);
		expect(screen.getByText('Hello ðŸ‘‹ world!')).toBeInTheDocument();
	});

	it('renders inline code when no className is provided', () => {
		const md = 'This has `inlineCode` without class.';
		const { container } = render(<Markdown>{md}</Markdown>);
		const codeEl = container.querySelector('code');
		expect(codeEl).toHaveTextContent('inlineCode');
	});

	it('renders ordered list properly', () => {
		render(<Markdown>{`1. First\n2. Second`}</Markdown>);
		expect(screen.getByText('First')).toBeInTheDocument();
		expect(screen.getByText('Second')).toBeInTheDocument();
	});

	it('renders preformatted block correctly', () => {
		const md = '```\npre block text\n```';
		render(<Markdown>{md}</Markdown>);
		expect(screen.getByText('pre block text')).toBeInTheDocument();
	});

	it('renders inline code using fallback', () => {
		render(<Markdown>{'Text with `inline`.'}</Markdown>);
		expect(screen.getByText('inline')).toBeInTheDocument();
	});

	it('renders inline code using fallback renderer', () => {
		const md = 'This has `fallbackCode`.';
		const { container } = render(<Markdown>{md}</Markdown>);
		const codeEl = container.querySelector('code');

		expect(codeEl).toBeInTheDocument();
		expect(codeEl?.textContent).toBe('fallbackCode');

		// Test the style to trigger coverage on all style lines (line 43 too)
		const style = window.getComputedStyle(codeEl!);
		expect(style.fontFamily).toMatch(/monospace/);
	});

	it('renders inline code with no className', () => {
		const md = 'This has `inlineCode` with no class.';
		const { container } = render(<Markdown>{md}</Markdown>);
		const code = container.querySelector('code');

		expect(code).toBeInTheDocument();
		expect(code?.textContent).toBe('inlineCode');
		expect(code).toHaveStyle({
			backgroundColor: '#f0f0f0',
		});
	});
});
