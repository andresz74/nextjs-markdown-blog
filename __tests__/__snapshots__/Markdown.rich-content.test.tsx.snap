// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown rich content rendering renders code block 1`] = `
<div>
  <pre>
    <div>
      <div
        style="background: rgb(40, 40, 40); color: rgb(171, 171, 171); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 2; hyphens: none; padding: 1em; margin: 0.5em 0px; overflow: auto; border-radius: 0.3em;"
      >
        <code
          class="language-ts"
          style="white-space: pre; background: rgb(40, 40, 40); color: rgb(171, 171, 171); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace; direction: ltr; text-align: left; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 2; hyphens: none;"
        >
          <span
            class="token"
            style="color: rgb(120, 120, 120);"
          >
            const
          </span>
          <span>
             x
          </span>
          <span
            class="token"
            style="color: rgb(97, 97, 97);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(121, 121, 121);"
          >
            number
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(97, 97, 97);"
          >
            =
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(102, 102, 102);"
          >
            42
          </span>
          <span
            class="token"
            style="color: rgb(171, 171, 171);"
          >
            ;
          </span>
          <span>
            

          </span>
        </code>
      </div>
    </div>
  </pre>
</div>
`;
